azurelinuxagent/ga/update.py

```diff
@@ -487,9 +488,44 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
         finally:
             self.last_incarnation = incarnation
 
+    @staticmethod
+    def __get_vmagent_upgrade_status(protocol, incarnation_changed):
+        """
+        This function gets the VMAgent upgrade status as per the last GoalState.
+        Returns: None if the last GS does not ask for requested version else VMAgentUpdateStatus
+        """
+        upgrade_status = None
+
+        try:
+            agent_manifests, _ = protocol.get_vmagent_manifests()
+            # Expectation here is that there will only be one manifest per family passed down from CRP.
+            # If there are multiple, we pick the first one
+            manifest = next(m for m in agent_manifests if m.family == conf.get_autoupdate_gafamily())
+            if manifest.is_requested_version_available:
+                if CURRENT_VERSION == manifest.requested_version:
+                    status = VMAgentUpdateStatuses.Success
+                    code = 0
+                else:
+                    status = VMAgentUpdateStatuses.Error
+                    code = -1
+                upgrade_status = VMAgentUpdateStatus(expected_version=manifest.version, status=status, code=code)
+        except StopIteration:
+            logger.verbose("No matching manifest found for family: {0}".format(conf.get_autoupdate_gafamily()))
+        except Exception as error:
+            if incarnation_changed:
+                err_msg = "[This error will only be logged once per incarnation] " \
+                          "Ran into error when trying to fetch updateStatus for the agent, skipping reporting update satus. Error: {0}".format(
+                           textutil.format_exception(error))
+                logger.warn(err_msg)
+                add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=err_msg, log_event=False)
+
+        return upgrade_status
+
     def _report_status(self, exthandlers_handler, incarnation_changed):
+        vm_agent_upgrade_status = self.__get_vmagent_upgrade_status(exthandlers_handler.protocol, incarnation_changed)
```

Comment:
should this be disabled in production until the feature is stable?

Reply:
This would only be reported if CRP includes a requested version in GS. This does no harm to the agent or CRP since its just adding to the status blob. Even if by a bug CRP sends us a requested version in GS before the feature is stable and we end up adding the `updateStatus` in the status blob, it wont do us any harm. It'll just be ignored.

Did you have any other use case in mind?

Reply:
where do we ensure that we report this only if the GS includes a requested version?

Reply:
The check for this is in Line 512 of update.py - 

```python
if manifest.is_requested_version_specified:
                if CURRENT_VERSION == manifest.requested_version:
                    status = VMAgentUpdateStatuses.Success
                    code = 0
                else:
                    status = VMAgentUpdateStatuses.Error
                    code = -1
                update_status = VMAgentUpdateStatus(expected_version=manifest.requested_version_string, status=status,
                                                    code=code)
```

In case the `manifest.is_requested_version_specified` condition is not satisfied, we return None. 

When we're preparing the status blob, we check if the `vm_status.vmAgent.update_status` is not None and if so, include the status in the blob else skip it (line 296 in wire.py) - 

```py
if ga_status.update_status is not None:
        v1_ga_status["updateStatus"] = get_ga_update_status_to_v1(ga_status.update_status)
```

Hope this helps!

Reply:
I still feel this should be completely disabled in production until the feature is stable. If at some point there is some issue in the feature and we have some agents pushing this status, who knows, that could be an issue.

Reply:
I don't see any big downside to it honestly. It's being controlled by CRP (the availability of `requested_version`) and we would upload status only if asked. On top of it, this status is only meant for RSM so it should be pretty invisible to the customers. 

On the flipside though, if I think about it, one case that comes to mind on when it can break is if CRP messes up and kind of couples the updateStatus with the extension status and fails extensions due to a failed upgrade. I guess we can go with a flagged approach and enable it by default when we have more confidence in the feature e2e. I'll also open a task to add a DCR scenario for it with the flag enabled and another for enabling it globally in the future once its ready.